src/lib.rs

```diff
@@ -0,0 +1,489 @@
+mod util;
+
+use cargo_metadata::PackageId;
+pub use config::*;
+pub use query::*;
+
+use check_release::run_check_release;
+use trustfall_rustdoc::{load_rustdoc, VersionedCrate};
+
+use itertools::Itertools;
+use rustdoc_cmd::RustdocCommand;
+use semver::Version;
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+
+/// Test a release for semver violations.
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct Check {
+    /// Which packages to analyze.
+    scope: Scope,
+    current: Rustdoc,
+    baseline: Rustdoc,
+    log_level: Option<log::Level>,
+}
+
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct Rustdoc {
+    source: RustdocSource,
+}
+
+impl Rustdoc {
+    /// Use an existing rustdoc file.
+    pub fn from_path(rustdoc_path: impl Into<PathBuf>) -> Self {
+        Self {
+            source: RustdocSource::Rustdoc(rustdoc_path.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the project root directory,
+    /// i.e. the directory containing the crate source.
+    /// It can be a workspace or a single package.
+    /// Same as [`Rustdoc::from_git_revision()`], but with the current git revision.
+    pub fn from_root(project_root: impl Into<PathBuf>) -> Self {
+        Self {
+            source: RustdocSource::Root(project_root.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the project at a given git revision.
+    pub fn from_git_revision(
+        project_root: impl Into<PathBuf>,
+        revision: impl Into<String>,
+    ) -> Self {
+        Self {
+            source: RustdocSource::Revision(project_root.into(), revision.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the largest-numbered non-yanked non-prerelease version
+    /// published to the cargo registry. If no such version, uses
+    /// the largest-numbered version including yanked and prerelease versions.
+    pub fn from_latest_version() -> Self {
+        Self {
+            source: RustdocSource::Version(None),
+        }
+    }
+
+    /// Generate the rustdoc file from a specific crate version.
+    pub fn from_version(version: impl Into<String>) -> Self {
+        Self {
+            source: RustdocSource::Version(Some(version.into())),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum RustdocSource {
+    /// Path to the Rustdoc json file.
+    /// Use this option when you have already generated the rustdoc file.
+    Rustdoc(PathBuf),
+    /// Project root directory, i.e. the directory containing the crate source.
+    /// It can be a workspace or a single package.
+    Root(PathBuf),
+    /// Project root directory and Git Revision.
+    Revision(PathBuf, String),
+    /// Version from cargo registry to lookup. E.g. "1.0.0".
+    /// If `None`, uses the largest-numbered non-yanked non-prerelease version
+    /// published to the cargo registry. If no such version, uses
+    /// the largest-numbered version including yanked and prerelease versions.
+    Version(Option<String>),
```

Comment:
Maybe it would be better to call it `Registry` instead of `Version`? It would make it more clear that it requires internet access to use it and from where this specific version is being taken from.

Reply:
What about `RegistryVersion`? 😄 
The user when seeing `Registry` with a string might think about `Registry("crates.io")` for example

Reply:
Naming is hard 😁
`VersionFromRegistry` is more verbose, but more clear maybe. 🤔

Reply:
I'm fine with either `RegistryVersion` or `VersionFromRegistry`. I'm not that worried about people thinking that they need to pass a version of a registry, but I'm willing to accept I may be overly optimistic if you think the other option is better.

Reply:
What do you both think about 8485cc5dfe51a91ba2e8fe2b071c631f24f67cfb?

Reply:
I think it's fine :)
I would not try to make the library perfect in this PR. Otherwise we will have to deal with too many merge conflicts.

I guess (and hope) that the library will be extracted in a separate crate, so we can still change the library API in future PRs.

Reply:
This PR is on my to-do list to review today. Sorry for the delay! Trustfall ended up in the spotlight (it was [GitHub's #​1 most-starred project](https://twitter.com/PredragGruevski/status/1623711641127747584) a couple of days ago) and I've been super busy over there for the last couple of days.

I know I probably just merged another PR that causes another merge conflict, and I'm happy to fix the mess of my own making 😅